### PR TITLE
REFACTOR: `paid` flag -> `PaymentResult` object

### PR DESCRIPTION
**Pull request by lollerfirst**
This PR is to deprecate the `paid` flag in `PaymentResponse` and `PaymentStatus` in favor of a more structured `PaymentResult`, that allows for more flexibility.
I have modified the backends to fill in this new field. The main difference is we DO NOT return `SETTLED` when paying an invoice unless we get an explicit confirmation, instead we return `PENDING` and then `melt` notices this and checks the payment status.
I have also adapted the code in `get_mint_quote` and `melt` to rely on the newly added `result` field.

### CODE REVIEW

File: cashu/lightning/base.py [hunk 1]
```diff
@@ -33,36 +34,79 @@ class InvoiceResponse(BaseModel):
...
+    FAILED = 0
+    SETTLED = 1
+    PENDING = 2
+
+    UNKNOWN = 3
+
+    def __str__(self):
+        return self.name
+
+    # We assume `None` is `FAILED`
```
Comment (prusnak) [inline cashu/lightning/base.py:47]: I think that `None` should be `PENDING`, no?

Comment (lollerfirst) [reply cashu/lightning/base.py:47]: Maybe you are right. But we also used `None` to signal what now would be `UNKNOWN`

Comment (lollerfirst) [reply cashu/lightning/base.py:47]: Shouldn't matter too much, though. One shouldn't be using that. I only ever created it to use it in `FakeWallet`

Comment (prusnak) [reply cashu/lightning/base.py:47]: Then I suggest to change this code to assume that None is PENDING.

That way it is consistent with rest of the old and new code.

**Comment by lollerfirst** [comment]
I am having trouble with LNBits.

The code for `get_invoice_status` reads a response field `paid` that isn't event mentioned in the docs (https://demo.lnbits.com/docs#/Payments/Payment_List_api_v1_payments_get)

File: cashu/lightning/base.py [hunk 2]
```diff
@@ -33,36 +34,81 @@ class InvoiceResponse(BaseModel):
...
+        if paid is None:
+            return cls.PENDING
+        elif not paid:
+            return cls.FAILED
+        elif paid:
+            return cls.SETTLED
 
 class PaymentResponse(BaseModel):
+    result: PaymentResult
     ok: Optional[bool] = None  # True: paid, False: failed, None: pending or unknown
```
Comment (callebtc) [inline cashu/lightning/base.py:59]:
```suggestion
```

I suggest we remove all legacy flags

File: cashu/lightning/base.py [hunk 3]
```diff
@@ -33,36 +34,81 @@ class InvoiceResponse(BaseModel):
...
+
+    # We assume `None` is `PENDING`
+    @classmethod
+    def from_paid_flag(cls, paid: Optional[bool]):
+        if paid is None:
+            return cls.PENDING
+        elif not paid:
+            return cls.FAILED
+        elif paid:
+            return cls.SETTLED
```
Comment (callebtc) [inline cashu/lightning/base.py:55]:
```suggestion
```

We can get rid of the old `paid` flag altogether.

File: cashu/lightning/base.py [hunk 4]
```diff
@@ -33,36 +34,81 @@ class InvoiceResponse(BaseModel):
...
+    def failed(self) -> bool:
+        return self.result == PaymentResult.FAILED
+
+    @property
+    def unknown(self) -> bool:
+        return self.result == PaymentResult.UNKNOWN
 
 class PaymentStatus(BaseModel):
+    result: PaymentResult
     paid: Optional[bool] = None
```
Comment (callebtc) [inline cashu/lightning/base.py:83]:
```suggestion
```

**Comment by callebtc** [comment]
This will be used to fix #614

File: tests/test_mint_lightning_blink.py [hunk 5]
```diff
@@ -135,7 +136,7 @@ async def test_blink_pay_invoice_failure():
         state=MeltQuoteState.unpaid,
     )
     payment = await blink.pay_invoice(quote, 1000)
-    assert not payment.ok
+    assert payment.result != PaymentResult.SETTLED
```
Comment (lollerfirst) [inline tests/test_mint_lightning_blink.py:139]:
```suggestion
    assert payment.settled
```

File: tests/test_mint_lightning_blink.py [hunk 6]
```diff
@@ -155,7 +156,7 @@ async def test_blink_get_invoice_status():
     }
     respx.post(blink.endpoint).mock(return_value=Response(200, json=mock_response))
     status = await blink.get_invoice_status("123")
-    assert status.paid
+    assert status.result == PaymentResult.SETTLED
```
Comment (lollerfirst) [inline tests/test_mint_lightning_blink.py:159]:
```suggestion
    assert status.settled
```

File: tests/test_wallet_regtest_mpp.py [hunk 7]
```diff
@@ -59,19 +59,24 @@ async def _mint_pay_mpp(invoice: str, amount: int, proofs: List[Proof]):
...
+        target=partial_pay_real_invoice, args=(invoice_payment_request, 32, 1)
+    )
 
     t1.start()
     t2.start()
     t1.join()
     t2.join()
 
-    assert wallet.balance <= 256 - 32
+    assert wallet.balance == 64
```
Comment (lollerfirst) [inline tests/test_wallet_regtest_mpp.py:79]:
```suggestion
    assert wallet.balance == 128 - 32
```
We have 128 in the wallet, but we only pay 32 from the wallet. The other 32 is paid from another node.

File: tests/test_mint_lightning_blink.py [hunk 8]
```diff
@@ -183,7 +184,7 @@ async def test_blink_get_payment_status():
     }
     respx.post(blink.endpoint).mock(return_value=Response(200, json=mock_response))
     status = await blink.get_payment_status(payment_request)
-    assert status.paid
+    assert status.result == PaymentResult.SETTLED
```
Comment (lollerfirst) [inline tests/test_mint_lightning_blink.py:187]:
```suggestion
    assert status.settled
```

File: tests/test_mint_lightning_blink.py [hunk 9]
```diff
@@ -102,7 +103,7 @@ async def test_blink_pay_invoice():
         state=MeltQuoteState.unpaid,
     )
     payment = await blink.pay_invoice(quote, 1000)
-    assert payment.ok
+    assert payment.result == PaymentResult.SETTLED
```
Comment (lollerfirst) [inline tests/test_mint_lightning_blink.py:106]:
```suggestion
    assert payment.settled
```

File: tests/test_mint_init.py [hunk 10]
```diff
@@ -376,4 +396,72 @@ async def test_startup_regtest_pending_quote_failure(wallet: Wallet, ledger: Led
...
+
+    # expect that no melt quote is pending
+    melt_quotes = await ledger.crud.get_all_melt_quotes_from_pending_proofs(
+        db=ledger.db
+    )
+    assert not melt_quotes
+
+    # expect that proofs are unspent
+    states = await ledger.db_read.get_proofs_states([p.Y for p in send_proofs])
+    assert all([s.unspent for s in states])
```
Comment (lollerfirst) [inline tests/test_mint_init.py:464]: That's awesome!

File: cashu/mint/ledger.py [hunk 11]
```diff
@@ -755,9 +758,8 @@ async def get_melt_quote(self, quote_id: str) -> MeltQuote:
             status: PaymentStatus = await self.backends[method][
                 unit
             ].get_payment_status(melt_quote.checking_id)
-            if status.paid:
+            if status.result == PaymentResult.SETTLED:
```
Comment (lollerfirst) [inline cashu/mint/ledger.py:761]:
```suggestion
            if status.settled:

```

**Review by lollerfirst — COMMENTED** [review]
Nice! Left some suggestions...

File: tests/test_mint_init.py (hunk 10 above)
Comment (lollerfirst) [reply tests/test_mint_init.py:464]: Do we only check this on startup?

**Comment by lollerfirst** [comment]
I think this is not needed anymore.

**Pull request closed without merging**